### PR TITLE
Add offline cache for downloaded files

### PR DESCRIPTION
**Pull request by futpib**
This allows `@reforged/maker-appimage` to work offline (or online faster) and allows for offline CI with guaranteed reproducible builds.

Highly inspired by `@electron/get`
https://github.com/electron/get/blob/4e0b055a7c8aa337215a8cd20880acc262d5a925/src/Cache.ts
https://github.com/electron/get/blob/4e0b055a7c8aa337215a8cd20880acc262d5a925/src/index.ts#L66C9-L66C9

You can check it by installing from here https://www.npmjs.com/package/@futpib/maker-appimage

### CODE REVIEW

File: makers/appimage/package.json [hunk 1]
```diff
@@ -23,6 +23,8 @@
   "dependencies": {
     "@electron-forge/maker-base": "^6.0.4",
     "@spacingbat3/lss": "^1.0.0",
+    "env-paths": "^2.2.1",
+    "fs-extra": "^10.1.0",
```
Comment (SpacingBat3) [inline makers/appimage/package.json:27]: I don't really like the idea of adding additional dependencies to the package, unless Node.js/EcmaScript lacks the implementation in their APIs (or we need the polyfill to support older versions of Node if the given feature is immature / experimental).

I also have a history of third-party dependencies breaking entire projects, so I rather avoid using stuff I could normally do myself and both of these deps don't really seem to be used in code in way I couldn't implement them just with the native API (either Node.js or EcmaScript).

File: makers/appimage/src/cache.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  public static getCacheDirectory(downloadUrl: string): string {
+    const parsedDownloadUrl = url.parse(downloadUrl);
+    const { search, hash, pathname, ...rest } = parsedDownloadUrl;
+    const strippedUrl = url.format({ ...rest, pathname: path.dirname(pathname ?? 'reforged') });
+
+    return crypto
+      .createHash('sha256')
+      .update(strippedUrl)
+      .digest('hex');
+  }
```
Comment (SpacingBat3) [inline makers/appimage/src/cache.ts:21]: This whole function doesn't really play well in our scenario. I'm pretty sure it doesn't work as intended with our current *mirrors* mechanism. If you want to generate a unique identifier for the cache asset filename, why not acknowledge instead all the information maker currently has about them, i.e. AppImageKit release and asset name (currently `"AppRun" | "runtime"`) ? Hashing the *stripped* URL doesn't sound as the best way to do it, plus we might consider of grouping cache into `{release}/{filename}` actually, as this stuff is expected to not change (with an exception to `type2runtime` / `continuous`, this should not be cached as it is expected to change frequently, unlike to full releases that shouldn't change at all).

File: makers/appimage/src/main.ts [hunk 3]
```diff
@@ -45,6 +46,27 @@ const nodeFetch = (() => {
...
+  if (!response.ok) {
+    throw new Error(`${filename} request failure (${response.status}: ${response.statusText}).`)
+  }
+
+  const fileBuffer = await response.arrayBuffer();
+
+  await cache.putFileInCache(url, filename, fileBuffer);
+
+  return fileBuffer;
+}
```
Comment (SpacingBat3) [inline makers/appimage/src/main.ts:68]: Caching should not be a forced and default feature IMO, it should be something optional and configurable with `cache` boolean in config (or env var, e.g. `REFORGED_CACHEDIR={bool/string}`) when implementing. Defaults should always be staying non-cached for now, so users who build stuff for themselves like once in their life stay with clean desktop, while CIs or any machines using Electron Forge frequently can opt-in for cache if they benefit from it.
For me, it takes around 3s of only packaging the AppImage, which is even shorter than Electron runs the `package` job (4s on my PC).
Another concept I had is to provide latest stable maker assets (AppRun, runtime) as separate NPM package that would be opt-dependency and could be used instead of cache. Eventually we could also add an entry to `config` to use custom `AppRun` / `runtime` from paths, I saw AppImageKit actually doing this.
Another issue is updating the cache, possibly the best idea would be to use cached files, but still update the cache in the background to ensure everything is fresh and up-to-date with the upstream. Because staying at `X` build of asset could actually make harder to reproduce the builds if these are no longer distributed at upstream for any reason (while this would be rather rare for AppImageKit to update already published stable release, it's *better to be safe than sorry*). This is also one of the reasons why distributing the `runtime` and `AppRun` sounds as a better solution + we would then rely on NPM to properly fetch the package and verify the checksums, leaving less to care for the maker.

File: makers/appimage/src/main.ts [hunk 4]
```diff
@@ -158,24 +182,12 @@ export default class MakerAppImage<C extends MakerAppImageConfig> extends MakerB
...
       /** Details about the AppImage runtime. */
       runtime: Object.freeze({
-        data: nodeFetch(parseMirror(`${remote.mirror.runtime}${remote.dir}/${remote.file}`,currentTag,"runtime"))
-          .then(response => {
-            if(response.ok)
-              return response.arrayBuffer()
-            else
-              throw new Error(`Runtime request failure (${response.status}: ${response.statusText}).`)
-          }),
+        data: cachedFetchArrayBuffer(cache, parseMirror(`${remote.mirror.runtime}${remote.dir}/${remote.file}`,currentTag,"runtime"), 'runtime'),
```
Comment (SpacingBat3) [inline makers/appimage/src/main.ts:185]: I guess, for sake of readability and to reduce duplicate parameters in functions, you could also use the `parseMirror` inside the `cachedFetchArrayBuffer` function. Also why not name it just `cachedFetch` or `getCached`,  I think the name itself is unnecessarily long here.

File: package.json [hunk 5]
```diff
@@ -20,7 +20,7 @@
     "@types/node": "^18.11.17",
     "@types/semver": "^7.3.13",
     "typedoc": "^0.24.6",
-    "typescript": "^5.0.2"
+    "typescript": "^5.1.6"
```
Comment (SpacingBat3) [inline package.json:23]:
```suggestion
    "typescript": "^5.0.2"
```
This is outside of the scope for this PR. Please do not bump any dependencies.

File: makers/appimage/src/cache.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import envPaths from 'env-paths';
+import * as fs from 'fs-extra';
+import * as path from 'path';
+import * as url from 'url';
+import * as crypto from 'crypto';
+
+const defaultCacheRoot = envPaths('reforged', {
+  suffix: '',
+}).cache;
```
Comment (SpacingBat3) [inline makers/appimage/src/cache.ts:9]: Maybe just use:
```suggestion
import { homedir } from 'node:os';

const defaultCacheRoot = `${ process.env["XDG_CACHE_HOME"] ?? `${ homedir() }/.cache` }/reforged`;
```
…as maker currently runs on Linux only + I would like to avoid `envPaths` dependency.

File: makers/appimage/src/cache.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import envPaths from 'env-paths';
```
Comment (SpacingBat3) [inline makers/appimage/src/cache.ts:1]:
```suggestion
```

File: makers/appimage/src/cache.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import * as fs from 'fs-extra';
+import * as path from 'path';
+import * as url from 'url';
+import * as crypto from 'crypto';
```
Comment (SpacingBat3) [inline makers/appimage/src/cache.ts:5]: I would avoid using `import * as {X} from '{Y}';` syntax, I think it is better to indicate what exactly APIs are used with `import { A, B, C } from 'Y'`. Also, I'm myself using double quotes for most of the strings and probably you should too for code consistency.

**Review by SpacingBat3 — CHANGES_REQUESTED** [review]
~~The overall code looks good and fits well the repository, but there are a few aspects that could be done better. I'm not really fan of messing around the dependencies that much through, I actually avoid adding here dependencies I don't find to be necessary for implementing the maker. I have limited trust for third-parties after history with malicious dependency (`colors.js`).~~

---

**PS:** I saw the link you've sent and I see quite a similarity between `@electron/get` implementation vs this. I don't think it is legal (see DMCA and/or other copyright-associated laws, depending on a country you live in) for you (and me therefore to accept such work and redistribute) to copy someone else's work and distribute it as yours without keeping with the license. I'll probably close this PR, unless I see the original author(s) claiming it can be reused here and relicensed.

**Comment by futpib** [comment]
@SpacingBat3 Did you intentionally close this or just wanted me to make changes according to you review?

**Comment by SpacingBat3** [comment]
> @SpacingBat3 Did you intentionally close this or just wanted me to make changes according to you review?

See [my comment above](https://github.com/SpacingBat3/ReForged/pull/11#pullrequestreview-1535548868) (**TL;DR** – I don't want to be DMCA'd):

> **PS:** I saw the link you've sent and I see quite a similarity between `@electron/get` implementation vs this. I don't think it is legal (see DMCA and/or other copyright-associated laws, depending on a country you live in) for you (and me therefore to accept such work and redistribute) to copy someone else's work and distribute it as yours without keeping with the license. I'll probably close this PR, unless I see the original author(s) claiming it can be reused here and relicensed.

The decision is still indefinitive, eg. if the original author directly sign offs the work to be commited on this ISC license.

**Comment by futpib** [comment]
Will you accept an alternative implementation (that will address all the comments you made here) in another PR if I make one?

**Comment by SpacingBat3** [comment]
> Will you accept an alternative implementation (that will address all the comments you made here) in another PR if I make one?

No, if it contains *stolen* implementation from `@electron/get` or any code that doesn't seems fully owned by you (I just can't accept stuff like this).

**Comment by futpib** [comment]
Will you accept an alternative implementation that contains no _stolen_ implementation from `@electron/get` nor any code that doesn't seems fully owned by me and also addresses other review comments you made here in another PR if I make one?

**Comment by SpacingBat3** [comment]
> Will you accept an alternative implementation that contains no _stolen_ implementation from `@electron/get` nor any code that doesn't seems fully owned by me in another PR if I make one?

Possibly. Also a lot of these are *suggestions*, i.e. stuff that just came to my mind and should be considered when developing this but not always implemented. I also personally thought about *env-based opt-in* cache implementation, e.g. if user provides `REFORGED_CACHE` env with an *accesible* path, this is where **all** makers on this monorepo will use the cache…
